main/src/library/Fixpoint/Solver.flix

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2021 Benjamin Dahse
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use Fixpoint/Ast.{Datalog, Constraint, HeadTerm, PredSym, Denotation};
+use Fixpoint/Ast.Datalog.Datalog;
+use Fixpoint/Ast.Constraint.Constraint;
+use Fixpoint/Ast.HeadPredicate.HeadAtom;
+use Fixpoint/Ast.PredSym.PredSym;
+use Fixpoint/Ram.RelSym;
+
+// TODO: The following tests have been commented and should be fixed an incommented.
+// Test.Dec.RelAlias
+// Test.Exp.Fixpoint.Fold
+// Test.Predicate.Filter
+// Test.Predicate.Guard
+// Test.Predicate.Union
+// Test.Term.Apply
+// Test.Term.Var.CapturedVar
+// Test.Term.Var.WildVar
+// fixpoint-computations-on-lattices
+// TestConstant
+// IFDS
+// IDE
+// SUOpt
+// FloydWarshall
+// TestIntroduction
+
+namespace Fixpoint {
+    pub def solve(d: Datalog[v]): Datalog[v] with Order[v], Eq[v], ToString[v] =
+        stratify(d) |>
+        solveWithStratification(d)
+
+    pub def solveWithStratification(d: Datalog[v], stf: Map[PredSym, Int32]): Datalog[v] with Order[v], Eq[v], ToString[v] =
+        let _ = notifyPreSolve(d, stf);
+        let model = (compile(d, stf) |>
+         queryStmt |>
+         lowerStmt |>
+         interpret |>
+         toAst) as & Pure;
+         let _ = notifyPostSolve(model);
+         model
+
+    pub def union(d1: Datalog[v], d2: Datalog[v]): Datalog[v] = match (d1, d2) {
+        case (Datalog(edb1, idb1), Datalog(edb2, idb2)) =>
+            Datalog(Array.append(edb1, edb2), Array.append(idb1, idb2)) as & Pure
+    }
+
+    pub def isSubsetOf(d1: Datalog[v], d2: Datalog[v]): Bool with Order[v], Eq[v], ToString[v] = match (d1, d2) {
+        case (Datalog(edb1, _), Datalog(edb2, _)) =>
+            use Fixpoint/Ram.toDenotation;
+            let db1 = toDatabase(edb1);
+            let db2 = predicatesOf(edb1) |> restrict(edb2) |> toDatabase;
+            Map.forall(relSym -> rel1 -> match toDenotation(relSym) {
+                case Denotation.Relational =>
+                    let rel2 = Map.getWithDefault(relSym, Map#{}, db2);
+                    Map.forall(t -> _ -> Map.memberOf(t, rel2), rel1)
+                case Denotation.Latticenal(_, leq, _, _) =>
+                    let rel2 = Map.getWithDefault(relSym, Map#{}, db2);
+                    Map.forall(k -> l1 -> match Map.get(k, rel2) {
+                        case Some(l2) => l1 `leq` l2
+                        case None => false
```

Comment:
What if the fact is bottom? Then its non-existence in the map would imply that they are actually equal?

Reply:
If we never store bottom explicitly, this should be fine. Of course that whole issue can be avoided by using bottom as a default value in the lookup.